File: app/lib/models/groupe_running.dart
import 'user.dart';

class GroupeRunning {
  final int id;
  final String nom;
  final String? niveau;
  final String? description;
  final User? responsable;
  final List<User> membres;

  GroupeRunning({
    required this.id,
    required this.nom,
    this.niveau,
    this.description,
    this.responsable,
    List<User>? membres,
  }) : membres = membres ?? const [];

  factory GroupeRunning.fromJson(Map<String, dynamic> json) {
    List<User> membresList = const [];
    if (json['membres'] != null && json['membres'] is List) {
      membresList = (json['membres'] as List)
          .map((e) => User.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return GroupeRunning(
      id: json['id'] as int,
      nom: json['nom'] as String,
      niveau: json['niveau'] as String?,
      description: json['description'] as String?,
      responsable: json['responsable'] != null
          ? User.fromJson(json['responsable'] as Map<String, dynamic>)
          : null,
      membres: membresList,
    );
  }
}

File: demo/src/main/java/com/example/demo/entity/GroupeRunning.java
package com.example.demo.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.BatchSize;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "groupes_running")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class GroupeRunning {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String nom;

    private String niveau;

    @Column(columnDefinition = "TEXT")
    private String description;

    /** Responsable du groupe (AdminGroupe) - relation 1-1. */
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "responsable_id", unique = true)
    private User responsable;

    /** Adhérents du groupe - relation N-N (contient). BatchSize réduit le N+1 quand plusieurs groupes sont chargés. */
    @BatchSize(size = 20)
    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "groupe_adherents",
            joinColumns = @JoinColumn(name = "groupe_id"),
            inverseJoinColumns = @JoinColumn(name = "adherent_id")
    )
    @Builder.Default
    private List<User> membres = new ArrayList<>();

    @OneToMany(mappedBy = "groupe")
    @Builder.Default
    @JsonIgnore
    private List<Evenement> evenements = new ArrayList<>();

    @OneToMany(mappedBy = "groupe")
    @Builder.Default
    @JsonIgnore
    private List<ProgrammeEntrainement> programmes = new ArrayList<>();
}
